ci: dump stacktraces in case of unit tests aborted or timed out

azat opened the pull request:
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### Changelog entry (a [user-readable short description](https://github.com/ClickHouse/ClickHouse/blob/master/docs/changelog_entry_guidelines.md) of the changes that goes into CHANGELOG.md):
ci: dump stacktraces in case of unit tests aborted or timed out

https://pastila.nl/?00e96eb0/196dff088258a597bd1e6154f2941015#0PHksfOuvvKFoWoL+HvbJg==

azat committed:
ci: dump stacktraces in case of unit tests aborted or timed out

azat commented on ci/praktika/result.py:
```diff
@@ -424,7 +424,10 @@ def from_gtest_run(cls, unit_tests_path, name="", with_log=False):
-                f"{unit_tests_path} --gtest_output='json:{ResultTranslator.GTEST_RESULT_FILE}'",
+                # With gdb we will capture stacktrace in case of abnormal termination and timeout (30 mins)
+                # 'catch syscall exit_group' is a hack to avoid exiting with non-zero exit code when process terminated successfully
+                # FIXME: do not run for sanitizers
```
@maxknv I need some help with CI magic, can you please help me resolve this todo?

azat replied:
Thanks!

maxknv reviewed (changes requested):
I'm trying to avoid clickhouse specifics inside praktika (./ci/praktika/*.py).
Lets check if the change works for you and I'll finish it later.
please don't merge like that.

maxknv committed:
move gdb handling from praktika to job script

maxknv committed:
fail ut on non zero exit code

azat commented on ci/praktika/result.py:
```diff
@@ -410,26 +411,38 @@ def create_new(cls, name, status, links=None, info="", results=None):
     @classmethod
-    def from_gtest_run(cls, unit_tests_path, name="", with_log=False):
+    def from_gtest_run(
+        cls, unit_tests_path, name="", with_log=False, command_prefix=""
```
```suggestion
        cls, unit_tests_path, name="", with_log=False, command_launcher=""
```

azat commented on ci/praktika/result.py:
```diff
@@ -481,7 +494,8 @@ def from_commands_run(
         res = True  # Track success/failure status
-        error_infos = []
+        info_lines = []
+        MAX_LINES_IN_INFO = 300
```
This is only for summary of the job, correct? I still can see complete logs/gdb logs?

azat replied:
Indeed, can it be the reason for https://github.com/ClickHouse/ClickHouse/issues/84140#issuecomment-3098218635 ?

azat commented on ci/praktika/result.py:
```diff
@@ -410,26 +411,38 @@ def create_new(cls, name, status, links=None, info="", results=None):
-                f"{unit_tests_path} --gtest_output='json:{ResultTranslator.GTEST_RESULT_FILE}'",
+                command,
             ],
-            with_log=with_log,
```
Now we do not pass through `with_log` argument? Also why not enable it by default?

maxknv committed:
comments

azat committed:
Fix data-race in gtest_workload_resource_manager

azat committed:
ci: enable stacktrace on timeout with gdb for non-asan

azat committed:
ci: add --privileged to allow attaching with gdb to unit tests

azat commented:
```
Unknown syscall name 'exit_group'.
```

azat committed:
ci: try to catch exit_group by syscall number

Since now CI reports [1]:

    Unknown syscall name 'exit_group'.

  [1]: https://s3.amazonaws.com/clickhouse-test-reports/PRs/84527/389b200810a6876a5e388344bb09c663e99864de//unit_tests_tsan/job.log

azat committed:
ci: verbose timeout for unit tests

azat committed:
ci: bt of all threads in unit tests

azat committed:
ci: cleanup gdb command for unit tests

azat committed:
ci: increase timeout for unit tests (just to trigger CI)

azat committed:
ci: replace catch syscall with break _Exit

azat committed:
ci: remove break on _Exit, No stack is not an error

maxknv commented:
> OK, tests that creates threads can be heavy, especially under sanitizers, I see that gdb eats a lot of CPU, so let's do it in another way, once it hits timeout (30min), we need to attach with gdb and dump stacktraces, @maxknv can you help with this?
> 
> UPD: the problem is that `catch syscall` by number is slow and we need to build gdb with `expat` to fix this (though interesting that `catch syscall exit_group` works locally using the fasttest image)

Can we keep the script simpler? send signal, attach stacktrace to the report if any

azat commented:
>Can we keep the script simpler? send signal, attach stacktrace to the report if any

That way we cannot record stacktrace in case of ABRT
